manticore/platforms/evm_world_state.py

```diff
@@ -0,0 +1,468 @@
+import logging
+from abc import ABC, abstractmethod
+from eth_typing import ChecksumAddress, URI
+from io import TextIOBase
+from typing import Dict, List, Optional, Set, Tuple, Union
+from urllib.parse import ParseResult, urlparse
+from web3 import Web3
+from ..core.smtlib import Array, ArrayVariable, BitVec, BitVecConstant, BitVecITE, BitVecZeroExtend, ConstraintSet
+from ..ethereum.state import State
+from ..exceptions import EthereumError
+
+logger = logging.getLogger(__name__)
+
+
+# sam.moelius: map records which (symbolic) offsets have been written to.  data holds the values
+# written.
+
+
+class Storage:
+    def __init__(self, constraints: ConstraintSet, address: int):
+        """
+        :param constraints: the ConstraintSet with which this Storage object is associated
+        :param address: the address that owns this storage
+        """
+        self._data = constraints.new_array(
+            index_bits=256,
+            value_bits=256,
+            name=f"STORAGE_DATA_{address:x}",
+            avoid_collisions=True,
+            # sam.moelius: The use of default here creates unnecessary if-then-elses.  See
+            # ArrayProxy.get in expression.py.
+            # default=0,
+        )
+
+    def __getitem__(self, offset: Union[int, BitVec]) -> Union[int, BitVec]:
+        return self.get(offset, 0)
+
+    def get(self, offset: Union[int, BitVec], default: Union[int, BitVec]) -> Union[int, BitVec]:
+        if not isinstance(default, BitVec):
+            default = BitVecConstant(256, default)
+        return self._data.get(offset, default)
+
+    def set(self, offset: Union[int, BitVec], value: Union[int, BitVec]):
+        self._data[offset] = value
+
+    def get_items(self) -> List[Tuple[Union[int, BitVec], Union[int, BitVec]]]:
+        items = []
+        array = self._data.array
+        while not isinstance(array, ArrayVariable):
+            items.append((array.index, array.value))
+            array = array.array
+        return items
+
+    def dump(self, stream: TextIOBase, state: State):
+        concrete_indexes = set()
+        for sindex in self._data.written:
+            concrete_indexes.add(state.solve_one(sindex, constrain=True))
+
+        for index in concrete_indexes:
+            stream.write(
+                f"storage[{index:x}] = {state.solve_one(self._data[index], constrain=True):x}\n"
+            )
+
+
+####################################################################################################
+
+
+class WorldState:
+    def new_storage(
+        self, constraints: ConstraintSet, address: int, items: Optional[Dict[int, int]] = None
+    ) -> Storage:
+        storage = Storage(constraints, address)
+        if items is not None:
+            for key, value in items.items():
+                storage.set(key, value)
+        return storage
+
+    @abstractmethod
+    def is_remote(self) -> bool:
```

Comment:
I'm guessing I'll be better the less we actually use this method.

Reply:
Yes, totally agree.